File: python/packages/spore_core/src/spore_core/output_schema.py
"""Output-schema delivery + enforcement (issue #139).

``ReactConfig.output`` was presence-validated by the
:class:`~spore_core.execution_registry.ExecutionRegistry` at startup but IGNORED
at runtime: the resolved schema was never delivered to the model and never
validated the terminal. This module is the hand-rolled validator + the frozen
literals that make delivery + enforcement deterministic and BYTE-IDENTICAL
across the four language ports (Rust is the reference at
``rust/crates/spore-core/src/output_schema.rs``).

Validator subset (matches the Ollama ``format`` channel): only ``type`` /
``required`` / ``properties`` / ``enum`` are honored — NO off-the-shelf
validator (they diverge across languages and would break byte-identical
fixtures). Anything else in the schema is ignored.

Evaluation order (first-match-wins — FROZEN):

1. Output does not parse as JSON → :data:`ERR_NOT_JSON`.
2. Root value's JSON type ≠ schema ``type`` → :func:`err_root_type`.
3. A ``required`` property absent (iterate ``required`` in ARRAY order) →
   :func:`err_missing_required`.
4. A present property's value type ≠ its subschema ``type`` →
   :func:`err_property_type`.
5. A present property's value ∉ its subschema ``enum`` →
   :func:`err_property_enum`.
6. Otherwise valid.

Determinism rules (parity-critical):

- ``properties`` are iterated in LEXICOGRAPHICALLY-SORTED key order for steps 4
  and 5 (NOT insertion order — Python's ``json`` does not sort by default).
- ``required`` is checked in its given ARRAY order; ``enum`` membership is
  order-independent but the ``{enum}`` rendering in the message is sorted.
- ``{enum}`` and ``{value}`` are rendered via
  :func:`~spore_core.model._canonicalize_json` (canonical compact key-sorted
  JSON), e.g. ``["error","ok"]``, ``"maybe"``. The enum SORT in the message is
  INTENTIONAL (determinism over author intent); the membership check itself is
  order-independent.
- JSON type names: ``object`` / ``array`` / ``string`` / ``number`` /
  ``integer`` / ``boolean`` / ``null``.
- ``integer`` is a SEMANTIC check: a number with no fractional part passes as
  ``integer`` (``42.0`` passes, ``42.5`` fails). Python ``bool`` is a subclass
  of ``int`` — a boolean must NOT validate as ``integer`` / ``number``.

These literals are HASH-LOAD-BEARING — the Rust/TS/Go ports reproduce these
exact bytes.
"""

from __future__ import annotations

import json
from typing import Any

from .model import _canonicalize_json


def feedback_message(error: str) -> str:
    """The validation-failure feedback appended as a USER-role message before a
    retry (#139, AC2). ``error`` is substituted with exactly one validator error
    string. Exact bytes (single spaces, no trailing newline) — HASH-LOAD-BEARING.
    """

    return (
        f"Your previous response did not match the required output schema. {error} "
        "Reply with only a JSON value that satisfies the schema."
    )


#: Validator error (step 1): the model's response did not parse as JSON.
ERR_NOT_JSON = "The response was not valid JSON."


def err_root_type(expected: str, actual: str) -> str:
    """Validator error (step 2): the root value's JSON type differs from the
    schema's ``type``."""

    return f'Expected type "{expected}" but found "{actual}".'


def err_missing_required(name: str) -> str:
    """Validator error (step 3): a ``required`` property is absent."""

    return f'Missing required property "{name}".'


def err_property_type(name: str, expected: str, actual: str) -> str:
    """Validator error (step 4): a present property's value type differs from its
    subschema ``type``."""

    return f'Property "{name}" should be type "{expected}" but found "{actual}".'


def err_property_enum(name: str, enum_json: str, value_json: str) -> str:
    """Validator error (step 5): a present property's value is not in its
    subschema ``enum``. ``enum_json`` and ``value_json`` are canonical compact
    key-sorted JSON."""

    return f'Property "{name}" must be one of {enum_json} but found {value_json}.'


def _json_type_name(v: Any) -> str:
    """The JSON type name of ``v`` (the JSON set; ``integer`` is reported as
    ``number`` here — the integer DISTINCTION is a schema-side semantic check,
    see :func:`_matches_type`). Python ``bool`` is a subclass of ``int`` and
    must be classified as ``boolean``, NOT ``number``."""

    if v is None:
        return "null"
    if isinstance(v, bool):
        return "boolean"
    if isinstance(v, (int, float)):
        return "number"
    if isinstance(v, str):
        return "string"
    if isinstance(v, list):
        return "array"
    if isinstance(v, dict):
        return "object"
    # Unreachable for ``json.loads`` output; defensive.
    return "null"


def _matches_type(v: Any, expected: str) -> bool:
    """Whether ``v`` satisfies the schema type name ``expected``. ``integer`` is
    a SEMANTIC check: a JSON number with no fractional part passes (``42.0``
    passes, ``42.5`` fails); a ``number`` schema accepts any JSON number. A
    boolean is NEVER a number/integer (Python ``bool`` ⊂ ``int`` is handled
    explicitly)."""

    if expected == "object":
        return isinstance(v, dict)
    if expected == "array":
        return isinstance(v, list)
    if expected == "string":
        return isinstance(v, str)
    if expected == "boolean":
        return isinstance(v, bool)
    if expected == "null":
        return v is None
    if expected == "number":
        # A boolean is not a number (bool ⊂ int in Python).
        return isinstance(v, (int, float)) and not isinstance(v, bool)
    if expected == "integer":
        if isinstance(v, bool):
            return False
        if isinstance(v, int):
            return True
        if isinstance(v, float):
            # Integral iff no fractional part (42.0 passes, 42.5 fails).
            return v.is_integer()
        return False
    # Unknown/absent type name: treat as satisfied (the subset only constrains
    # the seven JSON type names; anything else is not enforced).
    return True


def _schema_type(schema: Any) -> str | None:
    """The schema's ``type`` keyword as a string, if present."""

    if isinstance(schema, dict):
        t = schema.get("type")
        if isinstance(t, str):
            return t
    return None


def validate_output(response: str, schema: Any) -> str | None:
    """Validate the model's terminal ``response`` text against ``schema`` (#139).

    Returns ``None`` on a match, or the FIRST (lowest-numbered) FROZEN validator
    error string in the evaluation order. The subset honored is ``type`` /
    ``required`` / ``properties`` / ``enum``; everything else is ignored.
    Iteration order is fixed (see the module docs) so the returned error is
    byte-identical across languages.
    """

    # Step 1: parse.
    try:
        value: Any = json.loads(response.strip())
    except ValueError:
        return ERR_NOT_JSON

    # Step 2: root type.
    expected = _schema_type(schema)
    if expected is not None and not _matches_type(value, expected):
        return err_root_type(expected, _json_type_name(value))

    obj = value if isinstance(value, dict) else None
    props = schema.get("properties") if isinstance(schema, dict) else None
    props = props if isinstance(props, dict) else None

    # Step 3: required (ARRAY order). Only meaningful for an object value.
    required = schema.get("required") if isinstance(schema, dict) else None
    if isinstance(required, list) and obj is not None:
        for r in required:
            if isinstance(r, str) and r not in obj:
                return err_missing_required(r)

    # Steps 4 + 5 iterate ``properties`` in LEXICOGRAPHICALLY-SORTED key order.
    if obj is not None and props is not None:
        keys = sorted(props.keys())

        # Step 4: present-property type.
        for key in keys:
            subschema = props[key]
            sub_expected = _schema_type(subschema)
            if key in obj and sub_expected is not None:
                present = obj[key]
                if not _matches_type(present, sub_expected):
                    return err_property_type(key, sub_expected, _json_type_name(present))

        # Step 5: present-property enum membership.
        for key in keys:
            subschema = props[key]
            enum_arr = subschema.get("enum") if isinstance(subschema, dict) else None
            if key in obj and isinstance(enum_arr, list):
                present = obj[key]
                if not any(_json_equal(present, e) for e in enum_arr):
                    # Render the enum SORTED (determinism over author intent) and
                    # the value canonically — both via ``_canonicalize_json``.
                    sorted_members = sorted(enum_arr, key=_canonicalize_json)
                    enum_json = "[" + ",".join(_canonicalize_json(e) for e in sorted_members) + "]"
                    value_json = _canonicalize_json(present)
                    return err_property_enum(key, enum_json, value_json)

    # Step 6: valid.
    return None


def _json_equal(a: Any, b: Any) -> bool:
    """JSON value equality for enum membership. Mirrors serde_json's ``==``: a
    boolean is never equal to a number even though Python ``True == 1``."""

    if isinstance(a, bool) != isinstance(b, bool):
        return False
    return a == b


__all__ = [
    "ERR_NOT_JSON",
    "err_missing_required",
    "err_property_enum",
    "err_property_type",
    "err_root_type",
    "feedback_message",
    "validate_output",
]

File: go/spore-core/output_schema.go
// Output-schema delivery + enforcement (issue #139).
//
// ReactConfig.Output (a *SchemaRef) was presence-validated by the
// ExecutionRegistry at startup but IGNORED at runtime: the resolved schema was
// never delivered to the model and never validated the terminal. This file is
// the hand-rolled validator + the frozen literals that make delivery +
// enforcement deterministic and BYTE-IDENTICAL across the four language ports
// (Rust is the reference: rust/crates/spore-core/src/output_schema.rs).
//
// # Validator subset (matches the Ollama `format` channel)
//
// Only these JSON-schema keywords are honored — NO off-the-shelf validator
// (they diverge across languages and would break byte-identical fixtures):
// `type` / `required` / `properties` / `enum`. Anything else in the schema is
// ignored. validateOutput returns "" on a match or one of the FROZEN validator
// error strings below.
//
// # Evaluation order (first-match-wins — FROZEN)
//
//  1. Output does not parse as JSON → errNotJSON.
//  2. Root value's JSON type ≠ schema `type` → errRootType.
//  3. A `required` property absent (iterate `required` in ARRAY order) →
//     errMissingRequired.
//  4. A present property's value type ≠ its subschema `type` → errPropertyType.
//  5. A present property's value ∉ its subschema `enum` → errPropertyEnum.
//  6. Otherwise valid.
//
// # Determinism rules (parity-critical)
//
//   - `properties` are iterated in LEXICOGRAPHICALLY-SORTED key order for steps
//     4 and 5 (Go map iteration is randomized — keys MUST be sorted explicitly;
//     this is the highest-risk parity bug).
//   - `required` is checked in its given ARRAY order; `enum` membership is
//     order-independent but the `{enum}` rendering in the message is SORTED.
//   - `{enum}` and `{value}` are rendered via canonicalizeJSON (canonical
//     compact key-sorted JSON), e.g. ["error","ok"], "maybe". The enum SORT in
//     the message is INTENTIONAL (determinism over author intent); the
//     membership check itself is order-independent.
//   - JSON type names: object / array / string / number / integer / boolean /
//     null.
//   - `integer` is a SEMANTIC check, not a token check: a number with no
//     fractional part passes as `integer` (42.0 passes, 42.5 fails). Go
//     encoding/json decodes all JSON numbers to float64 by default, so the value
//     tree is decoded with UseNumber() and the integral check inspects the
//     json.Number's fractional part — NOT the token type.
//
// These literals are HASH-LOAD-BEARING — the four ports MUST reproduce these
// exact bytes.

package sporecore

import (
	"bytes"
	"encoding/json"
	"fmt"
	"math"
	"sort"
	"strings"
)

// feedbackMessage is the validation-failure feedback appended as a USER-role
// message before a retry (#139, AC2). The single argument is substituted with
// exactly one validator error string. Exact bytes (single spaces, no trailing
// newline) — HASH-LOAD-BEARING.
func feedbackMessage(validatorErr string) string {
	return "Your previous response did not match the required output schema. " +
		validatorErr +
		" Reply with only a JSON value that satisfies the schema."
}

// errNotJSON is the validator error (step 1): the model's response did not parse
// as JSON.
const errNotJSON = "The response was not valid JSON."

// errRootType is the validator error (step 2): the root value's JSON type
// differs from the schema's `type`.
func errRootType(expected, actual string) string {
	return fmt.Sprintf("Expected type %q but found %q.", expected, actual)
}

// errMissingRequired is the validator error (step 3): a `required` property is
// absent.
func errMissingRequired(name string) string {
	return fmt.Sprintf("Missing required property %q.", name)
}

// errPropertyType is the validator error (step 4): a present property's value
// type differs from its subschema `type`.
func errPropertyType(name, expected, actual string) string {
	return fmt.Sprintf("Property %q should be type %q but found %q.", name, expected, actual)
}

// errPropertyEnum is the validator error (step 5): a present property's value is
// not in its subschema `enum`. enumJSON and valueJSON are canonical compact
// key-sorted JSON.
func errPropertyEnum(name, enumJSON, valueJSON string) string {
	return fmt.Sprintf("Property %q must be one of %s but found %s.", name, enumJSON, valueJSON)
}

// jsonTypeName returns the JSON type name of v (the JSON set; an integer is
// reported as "number" here — the integer DISTINCTION is a schema-side semantic
// check, see matchesType). v is a value tree produced by encoding/json with
// UseNumber() (so numbers are json.Number, not float64).
func jsonTypeName(v any) string {
	switch v.(type) {
	case nil:
		return "null"
	case bool:
		return "boolean"
	case json.Number:
		return "number"
	case string:
		return "string"
	case []any:
		return "array"
	case map[string]any:
		return "object"
	default:
		return "null"
	}
}

// isIntegral reports whether the json.Number n has no fractional part (42.0 is
// integral, 42.5 is not). It parses the number as a float and checks the
// fractional part — NOT the token form. Mirrors Rust's f.fract() == 0.0. A
// number that fails to parse is not integral (defensive; UseNumber values always
// parse).
func isIntegral(n json.Number) bool {
	f, err := n.Float64()
	if err != nil {
		return false
	}
	return f == math.Trunc(f)
}

// matchesType reports whether v satisfies the schema type name expected. v is a
// UseNumber() value tree. `integer` is a SEMANTIC check: a JSON number with no
// fractional part passes (42.0 passes, 42.5 fails); a `number` schema accepts
// any JSON number (integer or not). An unknown/absent type name is treated as
// satisfied (the subset only constrains the seven JSON type names).
func matchesType(v any, expected string) bool {
	switch expected {
	case "object":
		_, ok := v.(map[string]any)
		return ok
	case "array":
		_, ok := v.([]any)
		return ok
	case "string":
		_, ok := v.(string)
		return ok
	case "boolean":
		_, ok := v.(bool)
		return ok
	case "null":
		return v == nil
	case "number":
		_, ok := v.(json.Number)
		return ok
	case "integer":
		n, ok := v.(json.Number)
		if !ok {
			return false
		}
		return isIntegral(n)
	default:
		// Unknown/absent type name: the subset only constrains the seven JSON
		// type names; anything else is not enforced.
		return true
	}
}

// schemaType returns the schema's `type` keyword as a string, and whether it was
// present and a string. schema is a UseNumber() value tree.
func schemaType(schema any) (string, bool) {
	obj, ok := schema.(map[string]any)
	if !ok {
		return "", false
	}
	t, ok := obj["type"].(string)
	return t, ok
}

// decodeJSONTree parses raw JSON into a value tree using json.Number for numbers
// (so the `integer` semantic check can inspect the fractional part). Returns
// (nil, false) on a parse error.
func decodeJSONTree(raw []byte) (any, bool) {
	dec := json.NewDecoder(bytes.NewReader(raw))
	dec.UseNumber()
	var tree any
	if err := dec.Decode(&tree); err != nil {
		return nil, false
	}
	return tree, true
}

// canonicalizeSchema returns the canonical compact key-sorted JSON encoding of a
// schema RawMessage (#139). The single cross-language source of truth for the
// schema bytes embedded in the directive seed AND reported on a violation, so
// they are identical across the four ports regardless of each language's map
// insertion order. A non-parseable schema (defensive) is returned verbatim as a
// trimmed string.
func canonicalizeSchema(schema json.RawMessage) string {
	tree, ok := decodeJSONTree([]byte(schema))
	if !ok {
		return strings.TrimSpace(string(schema))
	}
	return canonicalizeJSON(tree)
}

// validateOutput validates the model's terminal response text against schema
// (#139). schema is the resolved output schema (a json.RawMessage of canonical
// or author-order JSON).
//
// Returns "" on a match, or the FIRST (lowest-numbered) FROZEN validator error
// in the evaluation order. The subset honored is `type` / `required` /
// `properties` / `enum`; everything else in the schema is ignored. Iteration
// order is fixed (see the file docs) so the returned error is byte-identical
// across languages.
func validateOutput(response string, schema json.RawMessage) string {
	// Step 1: parse the response (trim surrounding whitespace, matching Rust's
	// serde_json::from_str(response.trim())).
	value, ok := decodeJSONTree([]byte(strings.TrimSpace(response)))
	if !ok {
		return errNotJSON
	}

	schemaTree, ok := decodeJSONTree([]byte(schema))
	if !ok {
		// A non-parseable schema constrains nothing (defensive; the registry
		// only stores valid JSON schemas).
		return ""
	}

	// Step 2: root type.
	if expected, present := schemaType(schemaTree); present {
		if !matchesType(value, expected) {
			return errRootType(expected, jsonTypeName(value))
		}
	}

	obj, _ := value.(map[string]any)
	schemaObj, _ := schemaTree.(map[string]any)

	// Step 3: required (ARRAY order). Only meaningful for an object value; a
	// non-object value already passed step 2 (or had no `type`), so a `required`
	// list applies to its members only when it IS an object.
	if schemaObj != nil {
		if required, ok := schemaObj["required"].([]any); ok && obj != nil {
			for _, r := range required {
				name, ok := r.(string)
				if !ok {
					continue
				}
				if _, present := obj[name]; !present {
					return errMissingRequired(name)
				}
			}
		}
	}

	// Steps 4 + 5 iterate `properties` in LEXICOGRAPHICALLY-SORTED key order.
	var props map[string]any
	if schemaObj != nil {
		props, _ = schemaObj["properties"].(map[string]any)
	}
	if obj != nil && props != nil {
		keys := make([]string, 0, len(props))
		for k := range props {
			keys = append(keys, k)
		}
		sort.Strings(keys)

		// Step 4: present-property type.
		for _, key := range keys {
			present, has := obj[key]
			if !has {
				continue
			}
			expected, ok := schemaType(props[key])
			if !ok {
				continue
			}
			if !matchesType(present, expected) {
				return errPropertyType(key, expected, jsonTypeName(present))
			}
		}

		// Step 5: present-property enum membership.
		for _, key := range keys {
			present, has := obj[key]
			if !has {
				continue
			}
			sub, ok := props[key].(map[string]any)
			if !ok {
				continue
			}
			enumArr, ok := sub["enum"].([]any)
			if !ok {
				continue
			}
			member := false
			presentCanon := canonicalizeJSON(present)
			for _, e := range enumArr {
				if canonicalizeJSON(e) == presentCanon {
					member = true
					break
				}
			}
			if !member {
				// Render the enum SORTED (determinism over author intent) and the
				// value canonically — both via canonicalizeJSON.
				rendered := make([]string, len(enumArr))
				for i, e := range enumArr {
					rendered[i] = canonicalizeJSON(e)
				}
				sort.Strings(rendered)
				enumJSON := "[" + strings.Join(rendered, ",") + "]"
				return errPropertyEnum(key, enumJSON, presentCanon)
			}
		}
	}

	// Step 6: valid.
	return ""
}
